libs/core/algorithms/include/hpx/parallel/container_algorithms/contains.hpp

```diff
@@ -0,0 +1,367 @@
+#include <iterator>
+#include <type_traits>
+
+namespace hpx::parallel { namespace detail {
+    struct contains : public algorithm<contains, bool>
+    {
+        constexpr contains() noexcept
+          : algorithm("contains")
+        {
+        }
+
+        template <typename ExPolicy, typename Iterator, typename Sentinel,
+            typename T, typename Proj>
+        static constexpr bool sequential(
+            ExPolicy, Iterator first, Sentinel last, const T& val, Proj&& proj)
+        {
+            return sequential_contains<std::decay<ExPolicy>>(
+                first, last, val, HPX_FORWARD(Proj, proj));
+        }
+
+        template <typename ExPolicy, typename Iterator, typename Sentinel,
+            typename T, typename Proj>
+        static util::detail::algorithm_result_t<ExPolicy, bool> parallel(
+            ExPolicy&& orgpolicy, Iterator first, Sentinel last, const T& val,
+            Proj&& proj)
+        {
+            using difference_type =
+                typename std::iterator_traits<Iterator>::difference_type;
+            difference_type count = detail::distance(first, last);
+            if (count <= 0)
+                return util::detail::algorithm_result<ExPolicy, bool>::get(
+                    false);
+
+            decltype(auto) policy = parallel::util::adapt_placement_mode(
+                HPX_FORWARD(ExPolicy, orgpolicy),
+                hpx::threads::thread_placement_hint::breadth_first);
+
+            using policy_type = std::decay_t<decltype(policy)>;
+            util::cancellation_token<> tok;
+            auto f1 = [val, tok, proj](Iterator first, std::size_t count) {
+                sequential_contains<policy_type>(first, val, count, tok, proj);
+                return tok.was_cancelled();
+            };
+
+            auto f2 = [](auto&& results) {
+                return std::any_of(hpx::util::begin(results),
+                    hpx::util::end(results),
+                    [](hpx::future<bool>& val) { return val.get(); });
+            };
+
+            return util::partitioner<policy_type, bool>::call(
+                HPX_FORWARD(decltype(policy), policy), first, count,
+                HPX_MOVE(f1), HPX_MOVE(f2));
+        }
+    };
+
+    template <typename ExPolicy, typename T, typename Enable = void>
+    struct contains_subrange_helper;
+
+    template <typename ExPolicy, typename T>
+    struct contains_subrange_helper<ExPolicy, T,
+        std::enable_if_t<!hpx::is_async_execution_policy_v<ExPolicy>>>
+    {
+        static bool get(T& itr, T& last)
+        {
+            return itr != last;
+        }
+    };
+
+    template <typename ExPolicy, typename T>
+    struct contains_subrange_helper<ExPolicy, T,
+        std::enable_if_t<hpx::is_async_execution_policy_v<ExPolicy>>>
+    {
+        static hpx::future<bool> get(hpx::future<T>& itr, T& last)
+        {
+            return itr.then(
+                [&last](hpx::future<T> it) { return it.get() != last; });
```

Comment:
@Pansysk75 Hi Panos, I'm not sure if this is correct, could you check this line of code I changed? I changed the helper function to return hpx::future, a continuation to the previous future, by using .then().

Reply:
That looks right, i think you might want to capture `last` by value and not by reference, as it is a local variable of the `parallel` function. Otherwise looks good!

Reply:
@Pansysk75 Just to make sure, should I also pass last by value for the functions in my contains_subrange_helper_class or only for the lambda?